pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -0,0 +1,181 @@
+#define _GNU_SOURCE
+
+#include <error.h>
+#include <errno.h>
+
+#define errorf(status, fmt, ...) \
+  error_at_line(status, errno, __FILE__, __LINE__, fmt, ##__VA_ARGS__)
+
+#include <dirent.h>
+#include <ftw.h>
+#define NOPENFD 10
+
+#include <sched.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sysexits.h>
+#include <unistd.h>
+
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+
+char *env_whitelist[] =
+  {"TERM", "DISPLAY", "HOME", "LANG", "XAUTHORITY",
+   "XDG_RUNTIME_DIR", "SSL_CERT_FILE",
+   "DBUS_SESSION_BUS_ADDRESS"};
+
+char **env_build(char *names[], size_t len)
+{
+  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
+
+  for (size_t i = 0; i < len; i++) {
+    if (env = getenv(names[i])) {
+      if (asprintf(ptr++, "%s=%s", names[i], env) == -1)
+	errorf(EX_OSERR, "asprintf");
+    }
+  }
+
+  *ptr = NULL;
+  return ret;
+}
+
+struct bind {
+  char *from;
+  char *to;
+};
+
+struct bind binds[] =
+  {{"/", "host"},
+   {"/proc", "proc"},
+   {"/sys", "sys"},
+   {"/nix", "nix"},
+   {"/tmp", "tmp"},
+   {"/var", "var"},
+   {"/run", "run"},
+   {"/dev", "dev"},
+   {"/home", "home"}};
+
+void bind(struct bind *bind)
+{
+  DIR *src = opendir(bind->from);
+
+  if (src) {
+    if (closedir(src) == -1)
+      errorf(EX_IOERR, "closedir");
+
+    if (mkdir(bind->to, 0755) == -1)
+      errorf(EX_IOERR, "mkdir");
+
+    if (mount(bind->from, bind->to, "bind", MS_BIND | MS_REC, NULL) == -1)
+      errorf(EX_OSERR, "mount");
+
+  } else {
+    // https://github.com/NixOS/nixpkgs/issues/31104
+    if (errno != ENOENT)
+      errorf(EX_OSERR, "opendir");
+  }
+}
+
+void spitf(char *path, char *fmt, ...)
+{
+  va_list args;
+  va_start(args, fmt);
+
+  FILE *f = fopen(path, "w");
+
+  if (f == NULL)
+    errorf(EX_IOERR, "spitf(%s): fopen", path);
+
+  if (vfprintf(f, fmt, args) == -1)
+    errorf(EX_IOERR, "spitf(%s): vfprintf", path);
+
+  fclose(f); // ignore for now, fails on /proc/self/{g,u}id_map
+}
+
+int nftw_rm(const char *path, const struct stat *sb, int type, struct FTW *ftw)
+{
+  if(remove(path) == -1)
+    errorf(EX_IOERR, "nftw_rm");
+}
+
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+int main(int argc, char *argv[])
+{
+  if (argc < 2) {
+    fprintf(stderr, "Usage: %s command [arguments...]\n"
+                    "Requires Linux kernel >= 3.19 with CONFIG_USER_NS.\n", argv[0]);
+    exit(EX_USAGE);
+  }
+
+  char *root = mkdtemp(strdup("/tmp/chrootenvXXXXXX"));
```

Comment:
nitpick: memory never freed.

Reply:
the following works for me:

```c
#include <stdio.h>
#include <stdlib.h>
int main() {
  char template[] = "/tmp/chrootenvXXXXXX";
  char *root = mkdtemp(template);
  if (!root) {
    perror("mkdtemp()");
  } else {
    printf("see %s\n", root);
  }
}
```

Reply:
Interesting, if I swap `template[]` with `*template` it results in segfault.

Reply:
Switched to this.

Reply:
`char *template = "string"` is a pointer to a constant which resides in the read only memory. `char template[] = "string"` is a mutable 7 bytes array (initialized with a string literal).